Source/JavaScriptCore/runtime/PropertyTable.cpp

```diff
@@ -136,12 +156,103 @@ void PropertyTable::destroy(JSCell* cell)
 
 PropertyTable::~PropertyTable()
 {
-    iterator end = this->end();
-    for (iterator iter = begin(); iter != end; ++iter)
-        iter->key->deref();
+    forEachProperty([&](auto& entry) {
+        entry.key()->deref();
+        return IterationStatus::Continue;
+    });
+    PropertyTableMalloc::free(bitwise_cast<void*>(m_indexVector & indexVectorMask));
+}
+
+void PropertyTable::seal()
+{
+    forEachPropertyMutable([&](auto& entry) {
+        entry.setAttributes(entry.attributes() | static_cast<unsigned>(PropertyAttribute::DontDelete));
+        return IterationStatus::Continue;
+    });
+}
+
+void PropertyTable::freeze()
+{
+    forEachPropertyMutable([&](auto& entry) {
+        if (!(entry.attributes() & PropertyAttribute::Accessor))
+            entry.setAttributes(entry.attributes() | static_cast<unsigned>(PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly));
+        else
+            entry.setAttributes(entry.attributes() | static_cast<unsigned>(PropertyAttribute::DontDelete));
+        return IterationStatus::Continue;
+    });
+}
+
+bool PropertyTable::isSealed() const
+{
+    bool result = true;
+    forEachProperty([&](const auto& entry) {
+        if ((entry.attributes() & PropertyAttribute::DontDelete) != static_cast<unsigned>(PropertyAttribute::DontDelete)) {
+            result = false;
+            return IterationStatus::Done;
+        }
+        return IterationStatus::Continue;
+    });
+    return result;
+}
 
-    PropertyTableMalloc::free(m_index);
+bool PropertyTable::isFrozen() const
+{
+    bool result = true;
+    forEachProperty([&](const auto& entry) {
+        if (!(entry.attributes() & PropertyAttribute::DontDelete)) {
+            result = false;
+            return IterationStatus::Done;
+        }
+        if (!(entry.attributes() & (PropertyAttribute::ReadOnly | PropertyAttribute::Accessor))) {
+            result = false;
+            return IterationStatus::Done;
+        }
+        return IterationStatus::Continue;
+    });
+    return result;
+}
 
+PropertyOffset PropertyTable::renumberPropertyOffsets(JSObject* object, unsigned inlineCapacity, Vector<JSValue>& values)
+{
+    ASSERT(values.size() == size());
+    unsigned i = 0;
+    PropertyOffset offset = invalidOffset;
+    forEachPropertyMutable([&](auto& entry) {
+        values[i] = object->getDirect(entry.offset());
+        offset = offsetForPropertyNumber(i, inlineCapacity);
+        entry.setOffset(offset);
+        ++i;
+        return IterationStatus::Continue;
+    });
+    clearDeletedOffsets();
+    return offset;
+}
+
+template<typename Functor>
+inline void PropertyTable::forEachPropertyMutable(const Functor& functor)
+{
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
```

Comment:
can this just use isCompact() and access the vector using the helper I suggested?